### ISSUE TITLE
adding Tox21 benchmark for none 3D models

### ISSUE THREAD

**Issue by thegodone** [aimat-lab/gcnn_keras#73]
I wonder if you can add the Tox21 benchmark case

**Comment by PatReis** [comment]
I added the Dataset and will run some training. Do you know how to treat the emtpy columns. I would just set it to 0 and then do multilabel training.

**Comment by thegodone** [comment]
We use this method for empty columns
def bce_loss(real, pred):
    real = tf.cast(real, dtype=tf.float32) 
    zlr = tf.zeros_like(real)
    nonenanreal = ~tf.math.is_nan(real)
    real = tf.where(nonenanreal, real, 0.)
    pred = tf.where(nonenanreal, pred, 0.)
    val = BCELoss(real, pred)
    return val

**Comment by thegodone** [comment]
Nice thanks

**Comment by PatReis** [comment]
I just tested with NaN in labels and added `BinaryCrossentropyNoNaN`, `BinaryAccuracyNoNaN` and `AUCNoNaN` loss and metrics. But the result is too good compared with literature benchmark. So I think they just replaced the missing values with 0's and trained as usual. If this is okay for you, I would do the same.

**Comment by PatReis** [comment]
I made it adjustable in Dataset init via `"remove_nan": True`

**Comment by thegodone** [comment]
this is not too good. in papers they compare AUC per groups (SR / NR) so it's 
![image](https://user-images.githubusercontent.com/1186658/188571792-0a1f350f-e9a6-44ea-9003-55c02b1b3395.png)

now look at the results of the best methods from 2014 cause on the true test file. I am not sure we have this file indeed...

![image](https://user-images.githubusercontent.com/1186658/188572011-a71bd37b-7b5c-487a-b755-46eefb462e7e.png)

forget the first line they use a larger dataset from 2019 (new assays) and compare to the historical benchmark from 2014.

**Comment by thegodone** [comment]
AMAZIZ is made by OCHEM

**Comment by PatReis** [comment]
Ah okay, but this is already with NaN removed, which is why I think this is okay. If I don't remove them I am at 0.95 AUC or something.

**Comment by thegodone** [comment]
0.95 AUC or ACC  ?

**Comment by PatReis** [comment]
You mean AttentiveFP for kgcnn models? Okay, I will test again.

**Comment by thegodone** [comment]
AttFP and DMPNN (aka chemprop)

**Comment by thegodone** [comment]
just look at your last results. You got them with NaNs or without Nans during  training ? cause this is good. 

special question: GAT is very interesting, I wonder if you can look at the LipschitzNorm (#76) that can be crucial for GAT to improve model for cheap in contrary to GATv2 (https://arxiv.org/abs/2103.04886)

**Comment by thegodone** [comment]
I see your point now on the metric i use a method target per target by removing nans. And use the mean and std of splits per targets aka columns

**Comment by PatReis** [comment]
Okay so if I set the empty classes to 0 (wihout NaNs) I get the result as in table.
DMPNN: Accuracy | AUC  --> 0.9423 ± 0.0015 | 0.8388 ± 0.0055

If I keep the NaNs and ignore them in loss and use the metrics below I get:
DMPNN: Accuracy | AUC  --> 0.8617 ± 0.0071 | 0.9381 ± 0.0016

The metrics should simply set the classes with NaN to 0 weight aka ignore them, cause NaN>0.5 will always register with False class value in keras.

```python
class BinaryAccuracyNoNaN(ks.metrics.BinaryAccuracy):

    def update_state(self, y_true, y_pred, sample_weight=None):
        is_not_nan = tf.cast(tf.logical_not(tf.math.is_nan(y_true)), y_pred.dtype)
        if sample_weight is not None:
            sample_weight *= is_not_nan
        else:
            sample_weight = is_not_nan
        return super(BinaryAccuracyNoNaN, self).update_state(y_true, y_pred, sample_weight=sample_weight)


class AUCNoNaN(ks.metrics.AUC):

    def update_state(self, y_true, y_pred, sample_weight=None):
        is_not_nan = tf.cast(tf.logical_not(tf.math.is_nan(y_true)), y_pred.dtype)
        if sample_weight is not None:
            sample_weight *= is_not_nan
        else:
            sample_weight = is_not_nan
        return super(AUCNoNaN, self).update_state(y_true, y_pred, sample_weight=sample_weight)
```

**Comment by thegodone** [comment]
I use this metric that is working properly for sigmoid MTL cases
```
def AccNaN(y_true, y_pred):
    """ NaN adapted Accuracy for sigmoid classification only """
    mask = tf.where(tf.math.is_nan(y_true),0.,1.)
    expt = tf.where(tf.math.is_nan(y_true),0., y_true) # We define 0 on NaN as placeholder
    pred = tf.where(y_pred>0.50, 1., 0.), 
    expt = tf.where(expt==1, 1., 0.)
    match = tf.cast(K.equal(expt,pred), dtype=tf.float32) # We compare and we convert to float
    # add control on den == 0!
    den = tf.math.reduce_sum(mask)
    den = tf.where(den==0.,1.,den)
    # We multiply by the mask 
    return tf.math.reduce_sum(mask*match)/den
```
![image](https://user-images.githubusercontent.com/1186658/188804361-4a722064-065c-4819-b87d-31a2476bb961.png)

**Comment by PatReis** [comment]
Okay, so your metric seems to do the same as the one I posted above. I just managed to swap AUC and Accuarcy names.
So it must be

Okay so if I set the empty classes to 0 (wihout NaNs) I get the result as in table.
DMPNN: Accuracy | AUC --> 0.9423 ± 0.0015 | 0.8388 ± 0.0055

If I keep the NaNs and ignore them in loss and use the metrics below I get:
DMPNN: Accuracy | AUC --> 0.9381 ± 0.0016 | 0.8617 ± 0.0071

So the Accuarcy more or less stayes the same but AUC get better if I correctly treat the NaNs.
So I think I will retrain with NaN is place.

**Comment by PatReis** [comment]
It is done now for Tox21, I will to the same for PROTEINS and may add CLinTox and SIDER

**Comment by thegodone** [comment]
You must keep the nans to get the more data as possible to learn. I think this explain the result you have in the second line here "If I keep the NaNs and ignore them in loss and use the metrics below I get"
DMPNN: Accuracy | AUC --> 0.9381 ± 0.0016 | 0.8617 ± 0.0071

**Comment by PatReis** [comment]
Well on the github page Chemprop report 0.851 ± 0.002 and that is not too far from our result.
However, I think the only remaining question is, how to treat multi-labels. Here, we just used normal Acc and AUC but with ignoring incomplete data/ NaNa. So all labels are essentiall flattened. 
You could on the other hand compute the Acc and AUC per class and average then. At least for the AUC this may be a larger difference. I can test this with keras by setting ``multi_label=True``.

**Comment by PatReis** [comment]
Okay I uploaded the training results for Tox21 with multi_label AUC keras metrics.

**Comment by PatReis** [comment]
I checked with deepchem `roc_auc_score` and except that the _NaN_-handling_ is not so easy, I get the same results as with keras AUC with multi-label (not multi-class) and shielding _NaN_,

I am closing this now. You can reopen if you have more questions about it...
